File: frontend/utils/subscription.ts
import { Subscription, SubscriptionFeatures, SubscriptionTier } from '@/api/types';

/**
 * Check if a subscription is currently active
 */
export function isSubscriptionActive(subscription: Subscription): boolean {
    if (!subscription) return false;
    
    if (subscription.status !== 'active' && 
        subscription.status !== 'trial' && 
        subscription.status !== 'canceled') {
        return false;
    }

    // Check if subscription has expired
    if (subscription.endDate) {
        const endDate = new Date(subscription.endDate);
        if (new Date() > endDate) {
            return false;
        }
    }

    return true;
}

/**
 * Check if subscription is Premium or Lifetime tier
 */
export function isPremiumTier(subscription: Subscription): boolean {
    if (!subscription) return false;
    return subscription.tier === 'premium' || subscription.tier === 'lifetime';
}

/**
 * Check if user has unlimited credits (Premium/Lifetime active)
 */
export function hasUnlimitedCredits(subscription: Subscription): boolean {
    return isSubscriptionActive(subscription) && isPremiumTier(subscription);
}

/**
 * Get the credit multiplier for a subscription
 * @returns 0 for unlimited, 1 for normal, 2 for 2x
 */
export function getCreditMultiplier(subscription: Subscription): number {
    if (!isSubscriptionActive(subscription)) {
        return 1.0;
    }

    switch (subscription.tier) {
        case 'basic':
            return 2.0;
        case 'premium':
        case 'lifetime':
            return 0.0; // 0 indicates unlimited
        default:
            return 1.0;
    }
}

/**
 * Get features available for a subscription
 */
export function getSubscriptionFeatures(subscription: Subscription | null): SubscriptionFeatures {
    if (!subscription || !isSubscriptionActive(subscription)) {
        return {
            unlimitedVoice: false,
            unlimitedNaturalLanguage: false,
            unlimitedGroups: false,
            unlimitedAnalytics: false,
            noAds: false,
            prioritySupport: false,
            creditMultiplier: 1.0,
        };
    }

    switch (subscription.tier) {
        case 'basic':
            return {
                unlimitedVoice: false,
                unlimitedNaturalLanguage: false,
                unlimitedGroups: false,
                unlimitedAnalytics: false,
                noAds: true,
                prioritySupport: false,
                creditMultiplier: 2.0,
            };
        case 'premium':
        case 'lifetime':
            return {
                unlimitedVoice: true,
                unlimitedNaturalLanguage: true,
                unlimitedGroups: true,
                unlimitedAnalytics: true,
                noAds: true,
                prioritySupport: true,
                creditMultiplier: 0.0, // Unlimited
            };
        default: // free
            return {
                unlimitedVoice: false,
                unlimitedNaturalLanguage: false,
                unlimitedGroups: false,
                unlimitedAnalytics: false,
                noAds: false,
                prioritySupport: false,
                creditMultiplier: 1.0,
            };
    }
}

/**
 * Format subscription tier for display
 */
export function formatTierName(tier: SubscriptionTier): string {
    switch (tier) {
        case 'free':
            return 'Free';
        case 'basic':
            return 'Basic';
        case 'premium':
            return 'Premium';
        case 'lifetime':
            return 'Lifetime';
        default:
            return 'Free';
    }
}

/**
 * Get tier color for UI elements
 */
export function getTierColor(tier: SubscriptionTier): string {
    switch (tier) {
        case 'basic':
            return '#3B82F6'; // Blue
        case 'premium':
            return '#8B5CF6'; // Purple
        case 'lifetime':
            return '#F59E0B'; // Gold
        default:
            return '#6B7280'; // Gray
    }
}

/**
 * Get price string for a tier
 */
export function getTierPrice(tier: SubscriptionTier): string {
    switch (tier) {
        case 'free':
            return '$0';
        case 'basic':
            return '$4.99/mo';
        case 'premium':
            return '$9.99/mo';
        case 'lifetime':
            return '$49.99';
        default:
            return '$0';
    }
}

/**
 * Check if user can use a specific feature
 */
export function canUseFeature(
    subscription: Subscription | null,
    feature: keyof SubscriptionFeatures
): boolean {
    const features = getSubscriptionFeatures(subscription);
    return features[feature] as boolean;
}

/**
 * Get days until subscription renews or expires
 */
export function getDaysUntilRenewal(subscription: Subscription): number | null {
    if (!subscription) return null;
    
    const targetDate = subscription.renewalDate || subscription.endDate;
    if (!targetDate) return null;

    const now = new Date();
    const target = new Date(targetDate);
    const diffTime = target.getTime() - now.getTime();
    const diffDays = Math.ceil(diffTime / (1000 * 60 * 60 * 24));
    
    return diffDays;
}

/**
 * Format renewal/expiration date for display
 */
export function formatRenewalDate(subscription: Subscription): string {
    const days = getDaysUntilRenewal(subscription);
    
    if (days === null) return '';
    if (days < 0) return 'Expired';
    if (days === 0) return 'Today';
    if (days === 1) return 'Tomorrow';
    if (days <= 7) return `In ${days} days`;
    
    const date = new Date(subscription.renewalDate || subscription.endDate!);
    return date.toLocaleDateString();
}

/**
 * Check if subscription is in grace period (canceled but still active)
 */
export function isInGracePeriod(subscription: Subscription): boolean {
    if (!subscription) return false;
    return subscription.status === 'canceled' && isSubscriptionActive(subscription);
}

/**
 * Get appropriate upgrade message based on current tier
 */
export function getUpgradeMessage(subscription: Subscription | null): string {
    if (!subscription || subscription.tier === 'free') {
        return 'Upgrade to Premium for unlimited access!';
    }
    if (subscription.tier === 'basic') {
        return 'Upgrade to Premium for unlimited everything!';
    }
    return 'You have the best plan!';
}

/**
 * Get subscription badge text
 */
export function getSubscriptionBadge(subscription: Subscription | null): string | null {
    if (!subscription || subscription.tier === 'free') return null;
    
    if (subscription.tier === 'lifetime') return '👑 Lifetime';
    if (subscription.tier === 'premium') return '⭐ Premium';
    if (subscription.tier === 'basic') return '✨ Basic';
    
    return null;
}

File: backend/internal/handlers/types/subscription.go
package types

import (
	"context"
	"time"

	"go.mongodb.org/mongo-driver/bson"
	"go.mongodb.org/mongo-driver/bson/primitive"
	"go.mongodb.org/mongo-driver/mongo"
)

// GetDefaultSubscription returns the default free tier subscription for new users
func GetDefaultSubscription() Subscription {
	now := time.Now()
	return Subscription{
		Tier:      TierFree,
		Status:    StatusActive,
		StartDate: now,
	}
}

// IsActive checks if the subscription is currently active
func (s *Subscription) IsActive() bool {
	if s.Status != StatusActive && s.Status != StatusTrial && s.Status != StatusCanceled {
		return false
	}

	// Check if subscription has expired
	if s.EndDate != nil && time.Now().After(*s.EndDate) {
		return false
	}

	return true
}

// IsPremiumTier checks if the subscription is Premium or Lifetime
func (s *Subscription) IsPremiumTier() bool {
	return s.Tier == TierPremium || s.Tier == TierLifetime
}

// HasUnlimitedCredits checks if the user has unlimited credits based on their subscription
func (s *Subscription) HasUnlimitedCredits() bool {
	return s.IsActive() && s.IsPremiumTier()
}

// GetCreditMultiplier returns the credit multiplier based on subscription tier
// Free: 1x, Basic: 2x, Premium/Lifetime: unlimited (represented as 0)
func (s *Subscription) GetCreditMultiplier() float64 {
	if !s.IsActive() {
		return 1.0
	}

	switch s.Tier {
	case TierBasic:
		return 2.0
	case TierPremium, TierLifetime:
		return 0.0 // 0 indicates unlimited
	default:
		return 1.0
	}
}

// SubscriptionFeatures represents what features are available for each tier
type SubscriptionFeatures struct {
	UnlimitedVoice           bool
	UnlimitedNaturalLanguage bool
	UnlimitedGroups          bool
	UnlimitedAnalytics       bool
	NoAds                    bool
	PrioritySupport          bool
	CreditMultiplier         float64
}

// GetFeatures returns the features available for this subscription
func (s *Subscription) GetFeatures() SubscriptionFeatures {
	if !s.IsActive() {
		return SubscriptionFeatures{
			CreditMultiplier: 1.0,
		}
	}

	switch s.Tier {
	case TierBasic:
		return SubscriptionFeatures{
			UnlimitedVoice:           false,
			UnlimitedNaturalLanguage: false,
			UnlimitedGroups:          false,
			UnlimitedAnalytics:       false,
			NoAds:                    true,
			PrioritySupport:          false,
			CreditMultiplier:         2.0,
		}
	case TierPremium, TierLifetime:
		return SubscriptionFeatures{
			UnlimitedVoice:           true,
			UnlimitedNaturalLanguage: true,
			UnlimitedGroups:          true,
			UnlimitedAnalytics:       true,
			NoAds:                    true,
			PrioritySupport:          true,
			CreditMultiplier:         0.0, // Unlimited
		}
	default: // TierFree
		return SubscriptionFeatures{
			CreditMultiplier: 1.0,
		}
	}
}

// UpdateSubscription updates a user's subscription in the database
func UpdateSubscription(ctx context.Context, userCollection *mongo.Collection, userID primitive.ObjectID, subscription Subscription) error {
	_, err := userCollection.UpdateOne(
		ctx,
		bson.M{"_id": userID},
		bson.M{"$set": bson.M{"subscription": subscription}},
	)
	return err
}

// GetUserSubscription retrieves a user's subscription from the database
func GetUserSubscription(ctx context.Context, userCollection *mongo.Collection, userID primitive.ObjectID) (*Subscription, error) {
	var user struct {
		Subscription Subscription `bson:"subscription"`
	}

	err := userCollection.FindOne(ctx, bson.M{"_id": userID}).Decode(&user)
	if err != nil {
		return nil, err
	}

	return &user.Subscription, nil
}

// UpgradeSubscription upgrades a user to a new subscription tier
func UpgradeSubscription(
	ctx context.Context,
	userCollection *mongo.Collection,
	userID primitive.ObjectID,
	tier SubscriptionTier,
	provider SubscriptionProvider,
	subscriptionID string,
) error {
	now := time.Now()

	// Calculate renewal date (30 days from now for monthly subscriptions)
	var renewalDate *time.Time
	if tier != TierLifetime {
		renewal := now.AddDate(0, 1, 0) // Add 1 month
		renewalDate = &renewal
	}

	subscription := Subscription{
		Tier:           tier,
		Status:         StatusActive,
		StartDate:      now,
		RenewalDate:    renewalDate,
		Provider:       provider,
		SubscriptionID: subscriptionID,
	}

	return UpdateSubscription(ctx, userCollection, userID, subscription)
}

// CancelSubscription cancels a user's subscription but keeps it active until end date
func CancelSubscription(ctx context.Context, userCollection *mongo.Collection, userID primitive.ObjectID) error {
	subscription, err := GetUserSubscription(ctx, userCollection, userID)
	if err != nil {
		return err
	}

	now := time.Now()
	subscription.Status = StatusCanceled
	subscription.CanceledAt = &now

	// Set end date to renewal date if not already set
	if subscription.EndDate == nil && subscription.RenewalDate != nil {
		subscription.EndDate = subscription.RenewalDate
	}

	return UpdateSubscription(ctx, userCollection, userID, *subscription)
}

// ExpireSubscription marks a subscription as expired and downgrades to free tier
func ExpireSubscription(ctx context.Context, userCollection *mongo.Collection, userID primitive.ObjectID) error {
	now := time.Now()
	subscription := Subscription{
		Tier:      TierFree,
		Status:    StatusActive,
		StartDate: now,
	}

	return UpdateSubscription(ctx, userCollection, userID, subscription)
}

// RenewSubscription renews a subscription for another billing period
func RenewSubscription(ctx context.Context, userCollection *mongo.Collection, userID primitive.ObjectID) error {
	subscription, err := GetUserSubscription(ctx, userCollection, userID)
	if err != nil {
		return err
	}

	now := time.Now()

	// Update renewal date to next month
	if subscription.Tier != TierLifetime {
		renewal := now.AddDate(0, 1, 0)
		subscription.RenewalDate = &renewal
	}

	// Clear canceled status and end date if renewing
	subscription.Status = StatusActive
	subscription.CanceledAt = nil
	subscription.EndDate = nil

	return UpdateSubscription(ctx, userCollection, userID, *subscription)
}

// CheckAndUpdateExpiredSubscriptions checks all subscriptions and expires those that should be expired
// This should be called periodically (e.g., daily cron job)
func CheckAndUpdateExpiredSubscriptions(ctx context.Context, userCollection *mongo.Collection) error {
	now := time.Now()

	// Find all subscriptions that should be expired
	filter := bson.M{
		"subscription.endDate": bson.M{"$lte": now},
		"subscription.status":  bson.M{"$in": []SubscriptionStatus{StatusActive, StatusTrial, StatusCanceled}},
	}

	// Update them to free tier
	_, err := userCollection.UpdateMany(
		ctx,
		filter,
		bson.M{
			"$set": bson.M{
				"subscription.tier":    TierFree,
				"subscription.status":  StatusExpired,
				"subscription.endDate": now,
			},
		},
	)

	return err
}
